Bt iso data fixes

Casper-Bonde-Bose opened the pull request:
A few fixes required for proper ISO data processing.

Fixes #31179.

Casper-Bonde-Bose committed:
Bluetooth: ISO: Add sequence number to ISO data packets

The sequence number is needed in the appliaction layer to detect lost
packets in the ISO stream in cases where the timestamp is not included.
(Sequence number is mandatory to include where timestamp is optional).
The API for the public metadata have been moved to a public header file.

As the size of the ISO meta data exceeds the default 4 octets net_buf
user_data the public ISO metadata have been moved into a seperate array.
The internal metadata is still stored in net_buf user_data.

This also fixed the user_data overflow on 32 bit systems, caused by
writing the ts into user_data on index 4 to 7, which is outside the 4
allocated bytes.

Signed-off-by: Casper Bonde <casper_bonde@bose.com>

Casper-Bonde-Bose committed:
Bluetooth: ISO: Speed up BT ISO data processing

For the native posix build the sleep calls used in tasks will stall
the zephyr instance which sets an upper limit on the data processing
interval to once every 20-30 millisecond.
This change reduces the duration of the sleep calls and increases the
TICKS_PER_SECOND to allow for shorter sleeps. This is needed to
support the data rates needed for LE Audio streaming. The rate is
tuned to support up to bidirectional 5ms ISO-intervals.
This change also increases the ISO buffer count from 1 to 5 to
allow for some buffering in the controller, which is needed for
gapless playback and/or use of burst number larger than 1.

Signed-off-by: Casper Bonde <casper_bonde@bose.com>

aescolar commented on boards/posix/native_posix/native_posix_defconfig:
```diff
@@ -4,6 +4,13 @@ CONFIG_SOC_POSIX=y
+# detection, hence data processing is at least two ticks per packet. To support
+# 5ms ISO interval bidirectional data the system shall never stall for more
+# than 1 ms.
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
```
This does not look like the change you would do in the default board configuration, but something you would do for your application..
Is there any reason to not do it in your application configuration file or application board configuration overlay?

Casper-Bonde-Bose replied:
As it is a general limiter for how many BT packets/events the native_posix build can handle per second, I see no reason not to make this change apply to all native_posix builds. It took quite a while for me to debug the low data rate, hence I'm sure it will save time for others in the future - it will also make the native_posix build behave more like a real system.

The implementation of userchan will be able to handle multiple packets if multiple packets are queued up in the BlueZ socket buffer. But as soon as you have emptied the socket buffer the system will sleep for one or two ticks before you read again. With 10 ms ticks this is more than two LC3 frames, hence will give an audio-gap for low latency audio. Also some controllers have a limited number of RX buffers for ISO data, hence will start to discard data.
As you say, the intention for this build is to have easy interaction with real host peripherals (such as audio), hence I think this change is needed, and I don't think I'm doing anything wrong. I think the wrong thing here is the way the native_posix implementation is done, where calls to read do not yield the task, which is why sleep is used in the first place - to busy-poll for data.
One of the next steps for ISO data is to include LC3 processing of ISO data, where you would need to process one packet every 7.5ms. The easiest way to actually get audio-data into Zephyr at the moment (where the audio subsystem is not in place) is to use a Linux build and pipe raw audio data in/out using a socket or file - which is why I would like to have the ISO data flow stable in the native posix build.

aescolar replied:
> I see no reason not to make this change apply to all native_posix builds.

native_posix is relatively rarely used with the BT host. It is not compiled in "real time mode" in most cases either.

> it will also make the native_posix build behave more like a real system

That was never the intention ;-). Native posix was born as a faster than real time test setup.

Anyhow, if you want to have this kconfig option default to a higher tick rate when BT_USER_CHAN is selected, that would be another matter (but @jhedberg would need to be ok with that).

>where calls to read do not yield the task

Call to the host read do not yield the task, because they are direct calls to the host OS, and have nothing to do with the Zephyr OS. There has been for a while a request to provide an API which can do what you would like #6044. But that is certainly not trivial to implement.

Casper-Bonde-Bose replied:
I see the tests fail due to 1000 ticks per second - hence I'll need to reduce to 500 or make it based on USERCHAN and/or the BT shell application.

Casper-Bonde-Bose replied:
>native_posix is relatively rarely used with the BT host.
I'll move this to the native_posix.conf overlay file for the relevant application then.

aescolar commented on subsys/shell/shell.c:
```diff
@@ -1131,7 +1131,8 @@ static void shell_log_process(const struct shell *shell)
 		if (shell->ctx->cmd_buff_len) {
-			k_sleep(K_MSEC(15));
+			k_yield();
+			k_sleep(K_MSEC(1)); /* Sleep do not seem to yield. */
```
k_sleep should yield if there is something to yield to. But if there is nothing to yield to nothing will seem to happen, except from time passing

Casper-Bonde-Bose replied:
Yes it should, but it does not seem to do that for the native posix build.

aescolar replied:
That should not be the case. So if you can create a case which reproduces it, I'd be very happy to see it.

Casper-Bonde-Bose replied:
I found that the native_posix build sets the system to 32 bit, hence user_data size was 4 - which caused caused the ISO subsystem to do memory corruption when writing 10 bytes of data into this 4 byte buffer. The result in a clean build is that the data flow stops after 5-10 ISO packets. When I added the calls to yield it fixed the data flow, hence I concluded k_sleep() did not yield, where the explanation is more likely that the linker rearranged some memory when rebuilding the files I changed.
After fixing the user_data size the calls to yield is no longer needed, and I'll remove it again.

aescolar reviewed (dismissed):
Commits' titles should be prefixed by the subsystem the commit targets: https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines

> `For the native posix build the sleep calls used i tasks will stall
the zephyr instance`

native_posix is not meant for hard real time interaction with host hardware. The real time support is only there to support some level of interactive testing, and testing with real host peripherals.
So if you find yourself needing to reconfigure it to awake every ms (as you are doing here), you are beyond the limit of what it ever was meant for. Even if native_posix tries to get rescheduled every ms, the host OS is going to schedule it with whatever priority at whatever tick it has. So you are going to see a lot of jitter in the scheduling.
If you want to interact with something sending a lot of data, you should, hopefully, mostly only need more buffering.

>`which sets an upper limit on the data processing speed to one packet every 20-30 millisecond.`

This sounds a bit funny. @jhedberg can the BT user channel not queue several packets at a time?

Casper-Bonde-Bose commented:
> Even if native_posix tries to get rescheduled every ms, the host OS is going to schedule it with whatever priority at whatever tick it has. So you are going to see a lot of jitter in the scheduling.

True, I think many people use this on a Linux PC, where the tick is typically 250 ticks per second. I do see some jitter, but it seems to be within one tick. My thinking of setting it to 1000 was that if multiple zephyr tasks was using sleep (at leas two does for the current ISO data test application), you could have zephyr execute 4 ticks for every host OS tick. When you start to add tasks to read audio data as well you would have more sleep calls.

Thalley commented on subsys/net/Kconfig:
```diff
@@ -15,7 +15,7 @@ if NET_BUF
 config NET_BUF_USER_DATA_SIZE
 	int "Size of user_data available in every network buffer"
-	default 8 if (BT || NET_TCP2) && 64BIT
+	default 10 if (BT || NET_TCP2) && 64BIT
```
Setting the default does not ensure correctness. I suggest adding a `depends on NET_BUF_USER_DATA_SIZE >= 0` in the BT_ISO Kconfig as well.

joerchan replied:
This must be broken on non 64BIT systems then. And instead if increasing the USER_DATA_SIZE I suggest to do what we do for HCI commands instead.

```
static struct cmd_data cmd_data[CONFIG_BT_HCI_CMD_COUNT];

#define cmd(buf) (&cmd_data[net_buf_id(buf)])
```

Casper-Bonde-Bose replied:
I simply added 2 octets to the already 8 octets of user data for ISO packets - and tracked the user-data size back to this define. It might not be the right one, but if it is not, I would appreciate a pointer to how the current 8 bytes of user-data is configured.
I don't understand this reference to hci commands - for HCI commands you save some context information about HCI commands that are being processed, where for ISO data you put the headers of the ISO data you receive from the controller into the user-data of the net_buf you pass on the application layer.

joerchan replied:
@Casper-Bonde-Bose Sorry for not explaining properly. My point was since user data of ISO was 8 bytes, and the default user data length for when `64BIT=n` then there must be an existing memory bug on 32-bit systems.
And since increasing the user-data size of the net bufs would affect the sizes of all the buffers in the system then that is quite an expensive.

To avoid this we have used a different approach for the HCI commands, by declaring an array of user_data, and using the
net_buf_id to index it.
Which should end up something like this then:
```
static struct iso_data iso_data[CONFIG_BT_ISO_RX_BUF_COUNT];
#define iso(buf) (&iso_data[net_buf_id(buf)])
```

Thalley replied:
Agreed; since this is used to hold the `struct iso_data` (i.e. ISO specific) I think a separate but "linked" array in iso.c is the better solution overall (will also be more future proof than incrementing the user_data size).

asmk-ot replied:
I ran into this issue when developing the iso adaptation layer (currently in draft), I added the following 
default 8 if ((BT || NET_TCP2) && 64BIT) || BT_ISO

This works for me while still using the userdata in the net_buf

Casper-Bonde-Bose replied:
@asmk-ot Yes, that would work for the current 8 bytes ISO user_data, but I extended to 10 bytes in this change. I think the issue people have with increasing this value is that all subsystems that use the net_buf system will bet a mem-penelty of 8 bytes per entry - hence the solution will be changed to use only the default 4 bytes of user_data and move the iso-specific data into a separate array.
By the way you will see all sorts of odd behavior on 32 bit systems when you actually start to receive ISO data. When you reach the last buffer in the pool you will overwrite what ever 4 bytes of memory the compiler/linker placed after the iso-pool array.

Casper-Bonde-Bose commented:
I uploaded a suggestion for the comment resolution. I'll need to rerun the tests when we have agreed on the proposed solution can be accepted.

Thalley commented on tests/bluetooth/shell/boards/native_posix.conf:
```diff
@@ -0,0 +1,12 @@
+# detection, hence data processing is at least two ticks per packet. To support
+# 5ms ISO interval bidirectional data the system shall never stall for more
+# than 1.5 ms in average.
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=500
```
Missing newline at end of file

Thalley replied:
@Casper-Bonde-Bose Still looks like it's missing the empty newline.

Casper-Bonde-Bose replied:
Yes, I have changed it locally, but are awaiting OK from Luiz on the API changes before I upload again - to avoid too many notifications to be send to people..

Thalley commented on include/bluetooth/iso.h:
```diff
@@ -362,6 +371,19 @@ int bt_iso_chan_disconnect(struct bt_iso_chan *chan);
  */
 int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf);
 
+/**
+ * @brief Get the metadata for a received ISO packet
+ *
+ * Received ISO data carry metadata such as ISO sequence  number and optionally
+ * a timestamp.
+ *
+ * @param buf pointer to a valid ISO buffer received through the recv callback.
+ *
+ * @return Pointer to the metadata for the buffer. The lifetime of the pointer
+ *         is linked to the lifetime of the net_buf.
+ */
+struct iso_meta_data *bt_iso_meta_data_get(struct net_buf *buf);
+
```
Hmm, since we are doing ISO API changes: Does it make more sense to just add a pointer to a `struct iso_meta_data` entry in the `struct bt_iso_chan_ops` `recv` callback instead?

Casper-Bonde-Bose replied:
@Vudentz do you agree? 
I like the idea, but wonder if there is a reason this was not already done?

Vudentz replied:
Or we don't strip the header which means the buffer would contain both the timestamp and sequence number as part of the data itself, that way we don't need to complicate either the user_data or reserver more space in the stack since the original buffer already contain the header.

Casper-Bonde-Bose replied:
@Vudentz I simply extended your implementation of stripping off the timestamp and store it in user_data. I actually like that the application do not need to write/call a parser for the header data. The solution with storing header data in a separate array do not use more memory than extending user-data or extending each pool (except on 64bit systems, where it will use a bit more).
As we already use the separate-array approach for HCI, could we agree to stick to this solution and provide the header in the call-back as Emil suggests?

Thalley replied:
Adding this to the receive callback will also following the convention of receive callbacks in bluetooth.h (i.e. an `info` object and the `data`).

Casper-Bonde-Bose replied:
Done

aescolar commented on subsys/shell/shell.c:
```diff
@@ -1131,7 +1131,7 @@ static void shell_log_process(const struct shell *shell)
 		 */
 		if (shell->ctx->cmd_buff_len) {
-			k_sleep(K_MSEC(15));
+			k_sleep(K_MSEC(1));
```
How come you need to do this change in the shell?

Casper-Bonde-Bose replied:
You are right, after fixing the memory corruption caused by net-buf user-data overflow (4 bytes allocated, 10 bytes used) the calls to sleep no longer block all execution in Zephyr, I'll revert it again.

Vudentz commented on subsys/bluetooth/host/iso.c:
```diff
@@ -28,6 +28,9 @@ NET_BUF_POOL_FIXED_DEFINE(iso_tx_pool, CONFIG_BT_ISO_TX_BUF_COUNT,
 
+static struct iso_meta_data iso_data[CONFIG_BT_ISO_RX_BUF_COUNT];
+#define iso_meta(buf) (&iso_data[net_buf_id(buf)])
+
```
Not sure I get this, why is this better than just increate the user_data? If we can't increase just for the ISO pool than perhaps that is the real problem here.

Casper-Bonde-Bose replied:
I agree, but it was suggested that increasing user_data was a bad idea, hence I changed to this approach as requested.

carlescufi commented:
@Casper-Bonde-Bose will you pick this up to drive it to completion?

Casper-Bonde-Bose commented:
> @Casper-Bonde-Bose will you pick this up to drive it to completion?
I was actually waiting for an agreement on what I'm allowed to do.
If I get no objections from @Vudentz , I'll go ahead and do the implementation @Thalley proposed.
Hence keep the `iso_meta_data iso_data[]` and include a pointer to an entry in the data call-back. And remove `bt_iso_meta_data_get()`

Casper-Bonde-Bose committed:
Bluetooth: ISO: Make it possible to setup unidirectional CIS

The current iso commands in the shell only supports setting up
a bidirectional stream or unidirectional for central role.
Adding rx/tx/rxtx option to iso listen command to allow for
peripheral side configuration of an iso connection.

Signed-off-by: Casper Bonde <casper_bonde@bose.com>

Casper-Bonde-Bose commented:
Rebased the change and made the agreed upon changes. Had to incorporate a bug-fix for setting up iso data path to be able to test the change.

Thalley commented on include/bluetooth/iso.h:
```diff
@@ -138,6 +138,15 @@ struct bt_iso_chan_path {
 
+/** @brief ISO Meta Data structure for received ISO packets. */
+struct bt_iso_recv_info {
+	/** ISO timestamp - valid only if the Bluetooth controller includes it */
+	uint32_t ts;
```
What is the value if the controller doens't include it? 0? 0xFFFFFFFF?

Casper-Bonde-Bose replied:
0 - I have update the text. (I have not changed this behavior in this PR)

Thalley commented on include/bluetooth/iso.h:
```diff
@@ -138,6 +138,15 @@ struct bt_iso_chan_path {
+	/** ISO timestamp - valid only if the Bluetooth controller includes it */
+	uint32_t ts;
+
+	/** ISO Pkt Seq no of the first fragment in the SDU */
```
```suggestion
	/** ISO packet sequence number of the first fragment in the SDU */
```

Casper-Bonde-Bose replied:
Done

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -104,6 +106,35 @@ static int cmd_listen(const struct shell *shell, size_t argc, char *argv[])
+	struct bt_conn *conns[1];
+	struct bt_iso_chan *chans[1];
+
+	if (!default_conn) {
+		shell_error(shell_ctx, "Not connected");
+		return 0;
+	}
+
+	conns[0] = default_conn;
```
I was wondering why we needed a check for `default_conn` for this one (it would make sense to me if the peripheral could setup this before any connection), but I guess we need to the set conns[0] pointer. 

Have you considered if it is possible to avoid doing this here?

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -104,6 +106,35 @@ static int cmd_listen(const struct shell *shell, size_t argc, char *argv[])
+	conns[0] = default_conn;
+	chans[0] = &iso_chan;
+
+	if (!strcmp("tx", argv[1])) {
+		chans[0]->qos->tx = &iso_tx_qos;
+		chans[0]->qos->rx = NULL;
+	} else if (!strcmp("rx", argv[1])) {
+		chans[0]->qos->tx = NULL;
+		chans[0]->qos->rx = &iso_rx_qos;
+	} else if (!strcmp("txrx", argv[1])) {
+		chans[0]->qos->tx = &iso_tx_qos;
+		chans[0]->qos->rx = &iso_rx_qos;
+	} else {
+		shell_error(shell_ctx, "Invalid argument - use tx, rx or txrx");
+	}
```
Hmm, I actually wonder if it would make sense to merge the peripheral and listen commands? The above check and assign of the connection pointer can't be done in listen, but the rest here seemingly can. Thoughts?

Casper-Bonde-Bose replied:
I can check if receiving the connection will reset any of the variables we setup as part the bind/peripheral command. But making it part of listen will limit the app to a single CIS. The code is written to be expandable to more connections - even though I do not know if that planned work.

Thalley reviewed (changes requested):
Two minor documentation comments, and then some questions regarding the peripheral command.

Thalley commented on tests/bluetooth/shell/boards/native_posix.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_NO_OPTIMIZATIONS=y
+# Allows for copying larger amount of data into the shell
+CONFIG_SHELL_BACKEND_SERIAL_RX_RING_BUFFER_SIZE=4096
+#Enable TINYCRYPT for native_posix; needed for CSIS
+CONFIG_BT_TINYCRYPT_ECC=y
```
I think this can be left out for now

Thalley reviewed (changes requested):
One minor final comment, otherwise LGTM

Thalley reviewed (approved):
LGTM

Casper-Bonde-Bose commented:
> This sounds a bit funny. @jhedberg can the BT user channel not queue several packets at a time?

@aescolar Just to conclude on this. Yes it can process multiple packets per tick, but if you have an ISO interval of 5ms and only have a few ISO buffers in the controller you will get gabs in the data if you only process data every 20-30ms. I did move the configuration of this to the BT-shell native posix test application configuration rather than the global configuration.

aescolar commented on tests/bluetooth/shell/boards/native_posix.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_LOG_MODE_IMMEDIATE=y
```
This should be the default in native_posix (no harm done having it here though)

Casper-Bonde-Bose replied:
Removed this line as suggested.

aescolar reviewed (commented):
>@aescolar Just to conclude on this. Yes it can process multiple packets per tick, but if you have an ISO interval of 5ms and only have a few ISO buffers in the controller you will get gabs in the data if you only process data every 20-30ms. I did move the configuration of this to the BT-shell native posix test application configuration rather than the global configuration.

Fair enough. I think at this point it is @jhedberg (and maybe also @Vudentz @joerchan) who should approve or request changes.

carlescufi commented:
@joerchan can you please re-review?

carlescufi commented:
@Casper-Bonde-Bose do you want this in for 2.6? if so please address the review comments

Casper-Bonde-Bose commented:
> @Casper-Bonde-Bose do you want this in for 2.6? if so please address the review comments.
@carlescufi There was a single optional comment - not something I expected would block the this patch, but now I made the requested change. Do I need to do something to indicate I have addressed all requested changes?

joerchan commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -474,7 +496,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(iso_cmds,
 		      "[latency] [sdu] [phy] [rtn]", cmd_bind, 1, 8),
 	SHELL_CMD_ARG(connect, NULL, "Connect ISO Channel", cmd_connect, 1, 0),
-	SHELL_CMD_ARG(listen, NULL, "[security level]", cmd_listen, 1, 1),
+	SHELL_CMD_ARG(listen, NULL, "[dir=tx,rx,txrx] [security level]", cmd_listen, 2, 1),
```
< > since this is a mandatory argument.

Casper-Bonde-Bose replied:
Thanks, Done.

Casper-Bonde-Bose commented:
> One thing that would be nice to add to the bt_iso_recv_info struct is also the packet flags (retrieved with the line `flags = bt_iso_pkt_flags(len);`), as that will tell the application if a packet is malformed or even lost.
> 
> Can be appended in this PR, or added in a new PR.

@Thalley I like this idea, and agree it would be a natural extension.

Thalley commented:
> > One thing that would be nice to add to the bt_iso_recv_info struct is also the packet flags (retrieved with the line `flags = bt_iso_pkt_flags(len);`), as that will tell the application if a packet is malformed or even lost.
> > Can be appended in this PR, or added in a new PR.
> 
> @Thalley I like this idea, and agree it would be a natural extension.

@Casper-Bonde-Bose : https://github.com/zephyrproject-rtos/zephyr/pull/35723
